components/hab/src/cli/hab/util.rs

```diff
@@ -18,24 +25,84 @@ use structopt::StructOpt;
-    /// Specify an alternate Builder endpoint. If not specified, the value will be taken from
-    /// the HAB_BLDR_URL environment variable if defined. (default: https://bldr.habitat.sh)
+    /// Specify an alternate Builder endpoint. If not specified, the value will be
+    /// taken from the HAB_BLDR_URL environment variable if defined. (default: https://bldr.habitat.sh)
     // TODO (DM): This should probably use `env` and `default_value`
     #[structopt(name = "BLDR_URL", short = "u", long = "url")]
-    bldr_url: Option<Url>,
+    pub value: Option<Url>,
+}
+
+#[derive(ConfigOpt, StructOpt, Deserialize, Serialize)]
+#[structopt(no_version)]
+#[configopt(derive(Serialize))]
+pub struct BldrOrigin {
+    /// The Builder origin name to target
+    #[structopt(name = "ORIGIN", short = "o", long = "origin")]
+    pub inner: Origin,
+}
+
+pub fn bldr_url_from_env_load_or_default() -> String {
+    bldr_url_from_env().unwrap_or_else(|| {
+                           match config::load() {
+                               Ok(config) => {
+                                   config.bldr_url
+                                         .unwrap_or_else(|| DEFAULT_BLDR_URL.to_string())
+                               }
+                               Err(e) => {
+                                   error!("Found a cli.toml but unable to load it. Resorting to \
+                                           default BLDR_URL: {}",
+                                          e);
+                                   DEFAULT_BLDR_URL.to_string()
+                               }
+                           }
+                       })
+}
+
+pub fn bldr_url_from_args_env_load_or_default(opt: Option<Url>) -> Result<Url, Error> {
+    if let Some(url) = opt {
+        Ok(url)
+    } else {
+        Url::parse(&bldr_url_from_env_load_or_default()).map_err(|e| {
+                                                            Error::InvalidUrl(e.to_string())
+                                                        })
+    }
+}
+
+pub fn bldr_auth_token_from_args_env_or_load(opt: Option<String>) -> Result<String, Error> {
+    if let Some(token) = opt {
+        Ok(token)
+    } else {
+        match henv::var(AUTH_TOKEN_ENVVAR) {
+            Ok(v) => Ok(v),
+            Err(_) => {
+                config::load()?.auth_token.ok_or_else(|| {
+                                              Error::ArgumentError("No auth token specified. \
+                                                                    Please check that you have \
+                                                                    specified a valid Personal \
+                                                                    Access Token with:  -z, \
+                                                                    --auth <AUTH_TOKEN>"
+                                                                                        .into())
+                                          })
+            }
```

Comment:
I think this could use `or_else`

Reply:
I believe this has to be `ok_or_else` since `config::load` is returning an option for `auth_token`. If I change to `or_else` clippy says:

```
error[E0308]: mismatched types
  --> components/hab/src/cli/hab/util.rs:90:17
   |
83 |   pub fn bldr_auth_token_from_args_env_or_load(opt: Option<String>) -> Result<String, Error> {
   |                                                                        --------------------- expected `std::result::Result<std::string::String, error::Error>` because of return type
...
90 | /                 config::load()?.auth_token.or_else(|| {
91 | |                                               Error::ArgumentError("No auth token specified. \
92 | |                                                                     Please check that you have \
93 | |                                                                     specified a valid Personal \
...  |
96 | |                                                                                         .into())
97 | |                                           })
   | |____________________________________________^ expected enum `std::result::Result`, found enum `std::option::Option`
   |
   = note: expected enum `std::result::Result<std::string::String, error::Error>`
              found enum `std::option::Option<std::string::String>`
```

I need to return an result so that the CLI marshaling can error if no auth token is found and one is required. We need to figure out how to handle this natively in the future (loading from args, ENV, OR cli.toml. See: https://github.com/habitat-sh/habitat/issues/7866

For now we need to return a Result so that we can error appropriately before calling the `start` function of the command:
```let auth_token = bldr_auth_token_from_args_env_or_load(r.auth_token.value)?;```

Reply:
Yeah, sorry I was not clear. I meant this part `match henv::var(AUTH_TOKEN_ENVVAR)` we could probably avoid the need to `match` on the `Ok` just to wrap it back up in `Ok`, but it would not really buy us that much.